moulinette/interfaces/api.py

```diff
@@ -27,6 +28,35 @@
 
 # API helpers ----------------------------------------------------------
 
+CSRF_TYPES = set(["text/plain",
+                  "application/x-www-form-urlencoded",
+                  "multipart/form-data"])
+
+
+def is_csrf():
+    """Checks is this is a CSRF request."""
+
+    if request.method != "POST":
+        return False
+    if request.content_type is None:
+        return True
+    content_type = request.content_type.lower().split(';')[0]
+    if content_type not in CSRF_TYPES:
+        return False
+
+    return request.headers.get("X-Requested-With") is None
```

Comment:
So, trying to understand this, very naive question : what prevents the attacker's code from adding this header to the request ? Aren't CSRF attacks designed to target specific pieces of software and therefore the attacker would implement this header ? Am I missing something ?

Reply:
The short answer is that the browser forbids the attacker's origin to set a custom HTTP Header unless the attacked origin explicitely allow that (with a `Access-Control-Allow-Headers` HTTP header).

The attack we are trying to prevent here (CSRF) is that the attacker code is executing from another origin (i.e. on another website). According to the fetch spec (https://fetch.spec.whatwg.org/), the attacker code is not allowed (by the browser) to add a custom HTTP Header (unless this is explicitely allowed by the target origin). It can only set `accept`, `accept-language`, `content-language`, `content-type` and few oddities.

Reply:
Okay that sounds good to me then ! Thanks for the explanation !